### PR TITLE
Log the actual wallet file version and no longer publicly expose the "version" record

### PR DESCRIPTION
**Pull request by achow101**
The wallet file version is stored in the "minversion" record, not the "version" record. However "version" is no longer used anywhere except to record the highest versioned client which has opened a wallet file (which is currently only used to check whether this was most recently opened by a 0.4.0 or 0.5.0rc1 client which had a broken wallet encryption implementation). Furthermore, "version" was logged to the debug.log which is confusing because it is not the actual wallet file version.

This PR changes it so that this confusion largely no longer exists. The wallet file version logging is changed to use "minversion" and reading and writing the "version" record is no longer publicly exposed to prevent potential confusion about whether the actual file version is being read or written. Lastly, in the one place it is actually used, the variable name is changed from nFileVersion to last_client to better reflect what that record actually represents.

### CODE REVIEW

**Comment by ryanofsky** [comment]
Just to clarify, is the only user-visible change here replacing:

```
nFileVersion = [version number of the newest wallet software that ever opened the wallet in the past]
```

with

```
Wallet File Version = [minimum wallet software version required to open the wallet file]
```

?

**Comment by achow101** [comment]
> Just to clarify, is the only user-visible change here replacing:

Yes.

**Review by jjjerrmmmy — APPROVED** [review]

**Comment by laanwj** [comment]
utACK 96a46f5bcd5433c3a4c5900ecc8be547e48f5f74

File: src/wallet/db.cpp [hunk 1]
```diff
@@ -587,7 +587,7 @@ BerkeleyBatch::BerkeleyBatch(BerkeleyDatabase& database, const char* pszMode, bo
             if (fCreate && !Exists(std::string("version"))) {
                 bool fTmp = fReadOnly;
                 fReadOnly = false;
-                WriteVersion(CLIENT_VERSION);
+                Write(std::string("version"), CLIENT_VERSION);
```
[the remark this answers is not in the document]
Comment (achow101) [reply src/wallet/db.cpp]: I'm not sure that this is redundant. It's making sure that the `version` field exists, whereas `WalletBatch::LoadWallet` doesn't necessarily do that.

File: src/wallet/walletdb.cpp [hunk 2]
```diff
@@ -376,12 +375,6 @@ ReadKeyValue(CWallet* pwallet, CDataStream& ssKey, CDataStream& ssValue,
 
             pwallet->LoadKeyPool(nIndex, keypool);
         }
-        else if (strType == "version")
-        {
-            ssValue >> wss.nFileVersion;
-            if (wss.nFileVersion == 10300)
-                wss.nFileVersion = 300;
```
[the remark this answers is not in the document]
Comment (achow101) [reply src/wallet/walletdb.cpp]: `nFileVersion` is being removed entirely, and all this `version` record did was fill it. At no point was `nFileVersion` being checked for anything less than 40000 so this if block is meaningless.

File: src/wallet/walletdb.cpp [hunk 3]
```diff
@@ -512,7 +505,12 @@ DBErrors WalletBatch::LoadWallet(CWallet* pwallet)
     if (result != DBErrors::LOAD_OK)
         return result;
 
-    pwallet->WalletLogPrintf("nFileVersion = %d\n", wss.nFileVersion);
+    // Last client version to open this wallet, was previously the file version number
+    int last_client = CLIENT_VERSION;
```
[the remark this answers is not in the document]
Comment (achow101) [reply src/wallet/walletdb.cpp]: I think I'll leave it as is.

**Comment by laanwj** [comment]
@meshcollider can you please take a look here

**Review by jb55 — CHANGES_REQUESTED** [review]
Concept ACK, I reviewed the changes in detail and confirmed they make sense logically. I have a few comments before I Approach ACK. Let me know what you think.

File: src/wallet/walletdb.cpp [hunk 4]
```diff
@@ -512,7 +512,8 @@ DBErrors WalletBatch::LoadWallet(CWallet* pwallet)
     if (result != DBErrors::LOAD_OK)
         return result;
 
-    pwallet->WalletLogPrintf("nFileVersion = %d\n", wss.nFileVersion);
+    int wallet_version = pwallet->GetVersion();
+    pwallet->WalletLogPrintf("Wallet File Version = %d\n", wallet_version > 0 ? wallet_version : wss.nFileVersion);
```
[the remark this answers is not in the document]
Comment (achow101) [reply src/wallet/walletdb.cpp]: Done

**Review by ryanofsky — APPROVED** [review]
utACK 35e60e790f2cd602d1bdd0be835d27f0ba37efa9. This code still pretty confusing, but a little simpler now. And the previous log statement was really misleading and useless compared to the new one here.

**Comment by jb55** [comment]
ACK 35e60e7, I compiled locally as a quick sanity check.

**Review by meshcollider — APPROVED** [review]
Looks good, thanks! utACK 35e60e790f2cd602d1bdd0be835d27f0ba37efa9

**Pull request merged**